design/ceph/object/ceph-cosi-driver.md

```diff
@@ -0,0 +1,140 @@
+
+#### Creating COSI Related CRDs
+
+There are five different kubernetes resources related to COSI. They are `Bucket`, `BucketAccess`, `BucketAccessClass`, `BucketClass` and `BucketClaim`. The user can create these resources using following `kubectl` command. All the examples will be added to deploy/examples/cosi directory in the Rook repository.
+
+```bash
+kubectl create -f deploy/examples/cosi/bucketclass.yaml
+kubectl create -f deploy/examples/cosi/bucketclaim.yaml
+kubectl create -f deploy/examples/cosi/bucketaccessclass.yaml
+kubectl create -f deploy/examples/cosi/bucketaccess.yaml
+```
+
+#### Consuming COSI Bucket
+
+The user needs to mount the secret as volume created by `BucketAccess` to the application pod. The user can use the secret to access the bucket by parsing the mounted file.
+
+```yaml
+spec:
+  containers:
+      volumeMounts:
+        - name: cosi-secrets
+          mountPath: /data/cosi
+  volumes:
+  - name: cosi-secrets
+    secret:
+      secretName: ba-secret
+```
+
+```bash
+# cat /data/cosi/bucket_info.json
+
+```json
+    {
+      apiVersion: "v1alpha1",
+      kind: "BucketInfo",
+      metadata: {
+          name: "ba-$uuid"
+      },
+      spec: {
+          bucketName: "ba-$uuid",
+          authenticationType: "KEY",
+          endpoint: "https://rook-ceph-my-store:443",
+          accessKeyID: "AKIAIOSFODNN7EXAMPLE",
+          accessSecretKey: "wJalrXUtnFEMI/K...",
+          region: "us-east-1",
+          protocols: [
+            "s3"
+          ]
+      }
+    }
+```
+
+#### Coexistence of COSI and lib-bucket-provisioner
+
+Currently the ceph object store provisioned via Object Bucket Claim (OBC). They both can coexist and can even use same backend bucket from ceph storage. No deployment/configuration changes are required to support both. The lib-bucket-provisioner is deprecated and eventually will be replaced by COSI when it becomes more and more stable. The CRDs used by both are different hence there is no conflicts between them.
+
+#### Transition from libbucket provisioner to COSI
+
+This applied to OBC with reclaim policy is `Retain` otherwise the bucket will be deleted when OBC is deleted. So no point in migrating the OBC with `Delete` reclaim policy.
+
+- First the user need to create a **COSI Bucket resource** pointing to the backend bucket.
+- Then user can create BucketAccessClass and BucketAccess using the COSI Bucket CRD.
+- Now the update application's credentials with BucketAccess secret, for OBC it was combination of secret and config map with keys word like AccessKey, SecretKey, Bucket, BucketHost etc. Here details in as JSON format in the [secret](#Consuming-COSI-Bucket).
+- Finally the user need to delete the existing OBC.
+
+### Points to remember
+
+#### Ceph COSI Driver Requirements
+
+- CephObjectStore should be deployed and running in the cluster
+- The credentials/endpoint for the CephObjectStore should be available by creating CephObjectStoreUser with proper permissions
```

Comment:
I think Rook can automatically create a `rook-ceph-object-store-<storename>-cosi-admin` user and store those credentials in a secret named `cosi-<storename>`.

BucketClasses can choose the object store used as a combo of namespace and storename, which directly translates to reading creds from `cosi-<storename>`. Optionally, to support manual deployment, this can also be taken to be namespace/<secretname>, and COSI can internally look for a secret called `<namespace>/<secret>` or `<namespace>/cosi-<secret>` to easily support Rook or user-defined.

Unfortunately, it's not possible to use globs in `resourceNames` in RBAC, but I think the benefits of keeping one secret per objectstore outweigh the privilege issues. Ultimately, the privilege problem is one present in k8s itself as pointed out by a user in the 
(discussion here)[https://github.com/kubernetes/kubernetes/issues/56582]. RBAC will have to allow Ceph COSI to read all secrets in the operator namespace.

Reply:
> RBAC will have to allow Ceph COSI to read all secrets in the operator namespace.

The secrets are in the same namespace as the object stores, right? So if the object stores are in a different namespace from the operator, we would need a binding to the secrets in each namespace where object stores are supported. In any case, agreed that the benefits of a single secret per object store outweights creating a single secret.

Reply:
Yes

Reply:
@BlaineEXE @travisn : I am trying to use the same secret generated by cephobjectstoreuser. The downside is the secret name will be in this format: `rook-ceph-object-user-<store-name>-cosi` But this controller needs to handle creating the secret/user etc since we already have in ceph-object-user-controller. What do you think about it?

Reply:
@travisn I understand cosi-<store-name> make more sense. But `CephObjectStoreUser` does not have a custom secret name option. I am trying to reuse existing code to make this controller simpler, otherwise we need to add copy logic from the ceph-object-user-controller to here. Another drawback of perms of the backend user will be exposed vis CRDs and it can be modified via CRDs.

Reply:
Sure, we can keep the pattern if it makes the implementation simpler

Reply:
I think this is the only detail that remains worth discussing, but it is small. We can leave it for a future change even after alpha.